core/src/main/java/amino/run/oms/migrationDecision/MigrationPredictor.java

```diff
@@ -0,0 +1,164 @@
+        // get node metrics
+        Map<InetSocketAddress, NodeMetric> currentNodeMetrics = nodeMetrics.get(replica.getHost());
+        Map<InetSocketAddress, NodeMetric> futureNodeMetric = nodeMetrics.get(futureKS);
+
+        long currentProcessingTime;
+        long currentByteExchanged;
+        long futureProcessingTime;
+        long currentNetworkFactorInProcessing;
+        long futureNetworkFactorInProcessing;
+        long timeOptimized;
+        InetSocketAddress kernelClient;
+        NodeMetric nodeMetric;
+        // evaluate processing time optimization for each kernel client
+        for (Map.Entry<InetSocketAddress, Long> processingTimeMetric :
+                processingTimeMetrics.entrySet()) {
+            kernelClient = processingTimeMetric.getKey();
+            currentProcessingTime = processingTimeMetric.getValue();
+            futureProcessingTime = (currentProcessingTime * futureKSCPU) / currentKSCPU;
+            // processingTimeMetric.getKey() = kernel client's kernelserver
+            currentByteExchanged = byteExchangedMetrics.get(kernelClient);
+
+            nodeMetric = currentNodeMetrics.get(kernelClient);
+            if (replica.getHost().equals(kernelClient)) {
+                currentNetworkFactorInProcessing = 0;
+            } else if (nodeMetric == null) {
+                logger.warning(
+                        String.format(
+                                "node metric not available between ks1 [%s] and ks2 [%s]",
+                                replica.getHost(), processingTimeMetric.getKey()));
+                continue;
+            } else {
+                currentNetworkFactorInProcessing =
+                        currentByteExchanged / nodeMetric.rate + nodeMetric.latency;
+            }
+
+            nodeMetric = futureNodeMetric.get(kernelClient);
+            if (futureKS.equals(kernelClient)) {
+                futureNetworkFactorInProcessing = 0;
+            } else if (nodeMetric == null) {
+                logger.warning(
+                        String.format(
+                                "node metric not available between ks1 [%s] and ks2 [%s]",
+                                futureKS, processingTimeMetric.getKey()));
+                continue;
+            } else {
+                futureNetworkFactorInProcessing =
+                        currentByteExchanged / nodeMetric.rate + nodeMetric.latency;
+            }
+
+            timeOptimized =
+                    (currentProcessingTime + currentNetworkFactorInProcessing)
+                            - (futureProcessingTime + futureNetworkFactorInProcessing);
+
+            // TODO Define SLA for Kernel client 
+            if (timeOptimized < 0) {
+                logger.info(
+                        String.format(
+                                "skipping kernel server [%s] as time optimization is negative",
+                                futureKS));
+                return 0;
```

Comment:
By this check , We are ensuring that because of migration to specific kernel server if any application faces negative optimisation (processing time degraded), we ignore migrating to that kernel server.